File: 2338_count_the_number_of_ideal_arrays.java
class Solution {
    static final int MOD = 1_000_000_007;
    static final int MAXLEN = 14;

    public int idealArrays(int n, int maxValue) {
        List<Integer>[] divisors = new ArrayList[maxValue + 1];
        for (int i = 1; i <= maxValue; i++) divisors[i] = new ArrayList<>();
        for (int i = 1; i <= maxValue; i++) {
            for (int j = i * 2; j <= maxValue; j += i) {
                divisors[j].add(i);
            }
        }

        long[][] dp = new long[MAXLEN + 1][maxValue + 1];
        for (int i = 1; i <= maxValue; i++) dp[1][i] = 1;

        for (int len = 2; len <= MAXLEN; len++) {
            for (int v = 1; v <= maxValue; v++) {
                for (int d : divisors[v]) {
                    dp[len][v] = (dp[len][v] + dp[len - 1][d]) % MOD;
                }
            }
        }

        long[] totalWays = new long[MAXLEN + 1];
        for (int len = 1; len <= MAXLEN; len++) {
            for (int v = 1; v <= maxValue; v++) {
                totalWays[len] = (totalWays[len] + dp[len][v]) % MOD;
            }
        }

        long[] fact = new long[n + 1], invFact = new long[n + 1];
        fact[0] = invFact[0] = 1;
        for (int i = 1; i <= n; i++) fact[i] = fact[i - 1] * i % MOD;
        for (int i = 1; i <= n; i++) invFact[i] = modInv(fact[i]);

        long res = 0;
        for (int len = 1; len <= Math.min(MAXLEN, n); len++) {
            long comb = fact[n - 1] * invFact[len - 1] % MOD * invFact[n - len] % MOD;
            res = (res + totalWays[len] * comb % MOD) % MOD;
        }

        return (int) res;
    }

    private long modInv(long x) {
        return pow(x, MOD - 2);
    }

    private long pow(long base, long exp) {
        long res = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) res = res * base % MOD;
            base = base * base % MOD;
            exp >>= 1;
        }
        return res;
    }
}

File: 2338_count_the_number_of_ideal_arrays.py
class Solution(object):
    def idealArrays(self, n, maxValue):
        """
        :type n: int
        :type maxValue: int
        :rtype: int
        """
        MOD = 10**9 + 7
        MAXLEN = 14  # log2(10^4)

        # Precompute all divisors
        divisors = [[] for _ in range(maxValue + 1)]
        for i in range(1, maxValue + 1):
            for j in range(i * 2, maxValue + 1, i):
                divisors[j].append(i)

        # dp[length][value]: number of sequences of length 'length' ending in 'value'
        dp = [[0] * (maxValue + 1) for _ in range(MAXLEN + 1)]
        for i in range(1, maxValue + 1):
            dp[1][i] = 1

        for length in range(2, MAXLEN + 1):
            for val in range(1, maxValue + 1):
                for d in divisors[val]:
                    dp[length][val] = (dp[length][val] + dp[length - 1][d]) % MOD

        # Total ideal arrays of each length
        total = [0] * (MAXLEN + 1)
        for length in range(1, MAXLEN + 1):
            total[length] = sum(dp[length]) % MOD

        # Precompute factorials and modular inverses
        fact = [1] * (n + 1)
        invfact = [1] * (n + 1)
        for i in range(1, n + 1):
            fact[i] = fact[i - 1] * i % MOD
        invfact[n] = pow(fact[n], MOD - 2, MOD)
        for i in range(n - 1, -1, -1):
            invfact[i] = invfact[i + 1] * (i + 1) % MOD

        # Combine counts with combinatorics: C(n-1, length-1)
        result = 0
        for length in range(1, min(MAXLEN, n) + 1):
            comb = fact[n - 1] * invfact[length - 1] % MOD * invfact[n - length] % MOD
            result = (result + total[length] * comb) % MOD

        return result
